tfexec/terraform.go

```diff
@@ -0,0 +1,606 @@
+package tfexec
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	tfjson "github.com/hashicorp/terraform-json"
+)
+
+type Terraform struct {
+	execPath    string
+	workingDir  string
+	execVersion string
+	env         []string
+	logger      *log.Logger
+}
+
+// NewTerraform returns a Terraform struct with default values for all fields.
+// If a blank execPath is supplied, NewTerraform will attempt to locate an
+// appropriate binary on the system PATH.
+func NewTerraform(workingDir string, execPath string) (*Terraform, error) {
+	var err error
+	if workingDir == "" {
+		return nil, fmt.Errorf("Terraform cannot be initialised with empty workdir")
+	}
+
+	if _, err := os.Stat(workingDir); err != nil {
+		return nil, fmt.Errorf("error initialising Terraform with workdir %s: %s", workingDir, err)
+	}
+
+	if execPath == "" {
+		execPath, err = FindTerraform()
+		if err != nil {
+			return nil, &ErrNoSuitableBinary{err: err}
+		}
+
+	}
+	tf := Terraform{
+		execPath:   execPath,
+		workingDir: workingDir,
+		env:        os.Environ(),
+		logger:     log.New(ioutil.Discard, "", 0),
+	}
+
+	execVersion, err := tf.version()
+	if err != nil {
+		return nil, &ErrNoSuitableBinary{err: fmt.Errorf("error running 'terraform version': %s", err)}
+	}
+
+	tf.execVersion = execVersion
+
+	return &tf, nil
+}
+
+func (tf *Terraform) SetEnv(env map[string]string) {
+	var tfenv []string
+
+	// always propagate CHECKPOINT_DISABLE env var unless it is
+	// explicitly overridden with tf.SetEnv
+	if _, ok := env["CHECKPOINT_DISABLE"]; !ok {
+		env["CHECKPOINT_DISABLE"] = os.Getenv("CHECKPOINT_DISABLE")
```

Comment:
This should probably be done at execution time to ensure its still picked up even if environments aren't set?